src/lasp_dependence_dag.erl

```diff
@@ -326,6 +322,35 @@ get_direct_edges(G, V1, V2) ->
         end
     end, digraph:out_edges(G, V1)).
 
+%% @doc Add a vertex only if it is top or bottom.
+%%
+%%      A top vertex is one that can't have parents.
+%%      Used as the source of updates.
+%%
+%%      A bottom vertex is one that can't have chidren.
+%%      Used as the destination of stream and reads.
+%%
+%%     We only add it if it isn't already present on the dag,
+%%     as adding the same vertex multiple times removes any
+%%     metadata (labels).
+%%
+-spec add_if(atom(), digraph:graph(), digraph:vertex()) -> ok.
+add_if(Setting, Dag, V) ->
+    Test = case Setting of
+        top -> fun is_top_vertex/1;
+        bottom -> fun is_bottom_vertex/1
+    end,
+    case Test(V) andalso (digraph:vertex(Dag, V) =:= false) of
+        true -> digraph:add_vertex(Dag, V);
+        _ -> ok
+    end.
+
+is_bottom_vertex(V) ->
+    lists:member(V, [stream, read, query]).
+
+is_top_vertex(bind) -> true;
+is_top_vertex(_) -> false.
```

Comment:
Yes, `update` works the same, it triggers a `bind` underneath. The dag will get notified that a given variable is being updated with a `bind`